### ISSUE TITLE
Error handling and dependencies in CLDR transformer code

### ISSUE THREAD

**Issue by sffc** [unicode-org/icu4x#574]
We decided in #558 to recommend String instead of more efficient types like Cow<str> and SmallString in the CLDR transformer to reflect that the CLDR transformer is an offline piece of code that should be optimized for readability and maintainability over performance.

I wanted to have a similar discussion about error handling. In #480, @gregtatum introduces some `panic!` into the CLDR transformer in a situation that would be an Error return value in library code. I can actually see some advantages to `panic!` in the CLDR transformer:

1. Easier to write
2. Stack traces tell the programmer exactly where the error occurred
3. Fewer return values need to be fallible

On the other hand, if we want to keep open the possibility that the CLDR transformer is run against untrusted CLDR JSON in production, then we should use the stack unraveling approach that we have in library code.

---

I also want to use this issue to follow up on the following comment from @gregtatum: https://github.com/unicode-org/icu4x/pull/541#discussion_r597788074

What should be the bar for adding new dependencies in provider_cldr?

**Comment by gregtatum** [comment]
Current error handling can be a bit annoying with lack of easy ways to display formatted error messages. If we do decide to return errors then I think we should only start doing that after #298 lands, since that will allow well written error messages.

For point 3, at the very least the `?` operator helps here make it not so onerous if we're always returning Results.

I'm not sure how exactly to address point 2.

**Comment by sffc** [comment]
2021-05-07: @Manishearth suggested using [eyre](https://crates.io/crates/eyre).  I also introduced anyhow in #704.  I agree with this approach and will follow up with a PR.

@nordzilla highlighted this talk about eyre: https://www.youtube.com/watch?v=rAF8mLI0naQ

**Comment by sffc** [comment]
@nordzilla Would you be interested in taking this on?  (migrating icu_provider_cldr to `eyre`)

**Comment by sffc** [comment]
So the problem with using `eyre` is that this is a library crate.  Perhaps thiserror is the tool for the job.  (Elsewhere we use displaydoc since it is `no_std`, but that's a non-goal of the CLDR provider.)

**Comment by sffc** [comment]
Everything uses DataError now. Nothing left to do.
